### DISCUSSION TITLE
Getting nan's from straightforward math operations.

### DISCUSSION THREAD

**Discussion by mark-at-hypothetic** [mitsuba-renderer/mitsuba3#893] [Q&A]
Hi! I'm trying to simply add to a tensor, but I'm getting nan's as a result.

Here's a simplified piece of code that gets me the nan's. Also, the number of nan's is variable each time I run it, but it's always divisible by 256. 

```
import drjit as dr
import mitsuba as mi
mi.set_variant('llvm_ad_rgb')


base_color_val = mi.TensorXf(np.ones((1,1,3)))   # TensorXf(shape=(1, 1, 3))
tex = mi.TensorXf(np.zeros((256,256,3)))            # TensorXf(shape=(256, 256, 3))
cc = tex + base_color_val                                      # TensorXf(shape=(256, 256, 3))

print(dr.sum(dr.isnan(cc))) --> out:  1792
print(dr.sum(dr.isfinite(cc))) --> out: 194816
print(256*256*3) --> out: 196608
```

but when I make base_color_val of shape (1,1,1) I get no nan's

```
base_color_val = mi.TensorXf(np.ones((1,1,3)))   # TensorXf(shape=(1, 1, 3))
tex = mi.TensorXf(np.zeros((256,256,3)))            # TensorXf(shape=(256, 256, 3))
cc = tex + base_color_val[:,:,0]                             # TensorXf(shape=(256, 256, 3))

print(dr.sum(dr.isnan(cc))) --> out: 0
print(dr.sum(dr.isfinite(cc))) --> out: 196608
print(256*256*3) --> out: 196608
```

obviously something is going wrong with slicing, but I guess it should've given an error instead of returning nan's?

Running 
mitsuba                   3.3.0
drjit                          0.4.2
ubuntu, python 3.10

**Comment by njroussel** [comment]
Hi @mark-at-hypothetic 

I'm unable to reproduce this.

You can "inspect" `cc` by converting it back to numpy for example. At least we might get some understanding of what is happening with its values.

As a side note, `drjit` has a `dr.ones` [operation](https://drjit.readthedocs.io/en/latest/reference.html#drjit.ones) and  a `dr.zeros` [operations](https://drjit.readthedocs.io/en/latest/reference.html#drjit.zeros) too.

**Reply by mark-at-hypothetic** [reply]
Hi! Thanks for getting back. dr.ones and zeros is good to know.
Inspecting the values with numpy, shows unexpected, extreme values. 
`3.567672e+09, 3.120646e-28, -6.238209e+10.`

When clipping, and substituting nan's, 
```
cc = cc.numpy()
cc[np.isnan(cc)] = 0.0
cc = np.clip(cc, 0.0, 1.0)
plt.imsave('cc.png', cc)
```
I get these textures, always a different pattern.
![cc](https://github.com/mitsuba-renderer/mitsuba3/assets/124044608/88a9aa11-db5e-4d8b-9538-4d7e3ed0dfb8)
![cc_2](https://github.com/mitsuba-renderer/mitsuba3/assets/124044608/a7d3ff62-5dfc-4abb-b15d-8a180443b7de)


Actually, your idea of using dr.ones and zeros is getting us a bit closer:
Using `base_color_val = dr.ones(dtype=dr.scalar.TensorXf, shape=(1,1,3)) ` (also for tex) gives the same error. But using CUDA (`dr.cuda.TensorXf`) for both the results are as expected.
`base_color_val = dr.ones(dtype=dr.cuda.TensorXf, shape=(1,1,3))`

Seems like a dr.jit bug

**Reply by njroussel** [reply]
I'm unable to reproduce this behavior at all. The fact that is only seems to happen with scalar and LLVM types leaves me to believe that there is some weird host system-level operation going on. The construction of the `TensorXf` object, is only a single `memcpy` normally. 

Feel free to open an issue.

**Reply by mark-at-hypothetic** [reply]
Strange. Thanks for looking into it. I'll open an issue

**Reply by tomas16** [reply]
I wonder if this is related to my issue #890. I'm using llvm_ad_rgb variant on macOS.

**Reply by njroussel** [reply]
Did either of you make any progress in identifying this issue?

**Reply by tomas16** [reply]
No, it'll be at least another month before I can continue working on this.
